source/Nevermore/Advanced/WriteTransaction.cs

```diff
@@ -50,13 +54,17 @@ public class WriteTransaction : ReadTransaction, IRelationalTransaction
 
         public async Task InsertAsync<TDocument>(TDocument document, InsertOptions options, CancellationToken cancellationToken = default) where TDocument : class
         {
-            var command = builder.PrepareInsert(new[] {document}, options);
-            await configuration.Hooks.BeforeInsertAsync(document, command.Mapping, this);
+            var commands = builder.PrepareInsert(new[] {document}, options);
```

Comment:
I actually don't know the answer to this: should we loop through _all_ commands  for the `BeforeInsertAsync` and `AfterInsertAsync` hooks? rather than processing each command one by one?

Actually - maybe we should only run the hook once. The mapping for the commands should all be the same right?

Reply:
Great question! I don't know the answer to this.

Reply:
> Actually - maybe we should only run the hook once. 

This makes sense to me. The fact that we might have one or more commands is an implementation detail. As a hook writer I would expect to get called once per document.

> The mapping for the commands should all be the same right?
We could make this assumption or decouple the code from the builder and retrieve the map directly from `configuration.DocumentMaps.Resolve()`.

Reply:
Ah it's resolved by `DataModificationQueryBuilder.GetMapping()`

Reply:
This has been updated now so hooks only get called once per document and the mapping has been pulled up to the `WriteTransaction` level.

https://github.com/OctopusDeploy/Nevermore/pull/153/commits/01bac4a56aaa403c90a91d79600fc47de6805d86